`FullscreenManager::exitFullscreen()` should not use `m_fullscreenElement`

nt1m opened the pull request:
#### 4624069466cfa7bc9cd5adc62c7caa0b489df70c
<pre>
`FullscreenManager::exitFullscreen()` should not use `m_fullscreenElement`
<a href="https://bugs.webkit.org/show_bug.cgi?id=287276">https://bugs.webkit.org/show_bug.cgi?id=287276</a>
<a href="https://rdar.apple.com/144403317">rdar://144403317</a>

Reviewed by Darin Adler and Brent Fulgham.

Reduce usage of `m_fullscreenElement`, so we can remove it.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::exitFullscreen):

Canonical link: <a href="https://commits.webkit.org/290722@main">https://commits.webkit.org/290722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69d1d05f73c8bfde80e3e99dd78f9da8e0f5fb03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90845 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10382 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95864 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41638 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92898 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10777 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69849 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27383 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8197 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82345 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50189 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36736 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40760 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37804 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97838 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18040 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13247 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78871 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18300 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78187 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78069 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19295 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22543 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21159 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18049 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23394 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17788 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21244 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -412,7 +412,8 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
         // there is a pending task in requestFullscreenForElement(). Cause it to cancel and fire an error
         // by clearing the pending fullscreen element.
-        if (!m_fullscreenElement && m_pendingFullscreenElement) {
+        RefPtr exitedFullscreenElement = fullscreenElement();
```
Why not `auto exitedFullscreenElement = protectedFullscreenElement()` ?

darinadler replied:
We generally like to write `RefPtr` instead of `auto` so you can see the lifetime while still not seeing a distractingly long type name, even though for types that are not pointers we are comfortable with `auto`.

cdumez replied:
Similarly, we use `protectedFoo()` functions to *avoid* local variables, not to initialize local variables. There is no perf benefit for initializing a local variable and it is less concise.

brentfulgham commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -428,10 +429,11 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
         // Notify the chrome of the new full screen element.
         if (mode == ExitMode::Resize)
-            page->chrome().client().exitFullScreenForElement(m_fullscreenElement.get());
+            page->chrome().client().exitFullScreenForElement(exitedFullscreenElement.get());
```
Nit: It seems like `exitFullscreenForElement(..)` should take a RefPtr, since it does various operations inside that might affect the life cycle of the element.

darinadler replied:
No, this is not how arguments in our safer C++ coding style work; the guidelines document explains this. @rniwa is the one who originally explained it to me.

brentfulgham reviewed (approved):
Looks good to me.

darinadler reviewed (approved):
I didn’t review this at first because I presumed it was a behavior-changing pull request that fixed a bug, with no test. But I see now that it’s a refactoring that is supposed to **not** change behavior.

nt1m committed:
`FullscreenManager::exitFullscreen()` should not use `m_fullscreenElement`
https://bugs.webkit.org/show_bug.cgi?id=287276
rdar://144403317

Reviewed by Darin Adler and Brent Fulgham.

Reduce usage of `m_fullscreenElement`, so we can remove it.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::exitFullscreen):

Canonical link: https://commits.webkit.org/290722@main
